### PR TITLE
Use patricia trie to OnHeapTrieBasedStringDictionary

### PR DESCRIPTION
**Pull request by jackjlli**

### CODE REVIEW

File: pinot-common/src/main/java/org/apache/pinot/common/config/IndexingConfig.java [hunk 1]
```diff
@@ -75,6 +75,9 @@
   @ConfigKey("onHeapDictionaryColumns")
   private List<String> _onHeapDictionaryColumns;
 
+  @ConfigKey("onHeapDictionaryColumns")
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/config/IndexingConfig.java]: Wrong name

File: pinot-common/src/main/java/org/apache/pinot/common/config/IndexingConfig.java [hunk 2]
```diff
@@ -179,6 +182,10 @@ public void setColumnMinMaxValueGeneratorMode(String columnMinMaxValueGeneratorM
     return _onHeapDictionaryColumns;
   }
 
+  public List<String> getOnHeapTrieBasedDictionaryColumns() {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/config/IndexingConfig.java]: Will this be a subset of onHeapDictionaryColumns?

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/OnHeapTrieBasedDictionary.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import org.apache.pinot.core.segment.memory.PinotDataBuffer;
+
+
+/**
+ * Abstract base class for all on-heap trie dictionary implementations.
+ *
+ */
+public abstract class OnHeapTrieBasedDictionary extends ImmutableDictionaryReader {
+
+  protected OnHeapTrieBasedDictionary(PinotDataBuffer dataBuffer, int length, int numBytesPerValue, byte paddingByte) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/OnHeapTrieBasedDictionary.java]: Do we need this constructor?

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/OnHeapTrieBasedStringDictionary.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+      return (id != null) ? id : Arrays.binarySearch(_unpaddedStrings, rawValue);
+    } else {
+      String paddedValue = padString((String) rawValue);
+      return Arrays.binarySearch(_paddedStrings, paddedValue);
+    }
+  }
+
+  @Override
+  public String get(int dictId) {
+    return _unpaddedStrings[dictId];
```
Comment (chenboat) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/OnHeapTrieBasedStringDictionary.java]: Do you need to check dictId is not out of bound?

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/OnHeapTrieBasedStringDictionary.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  }
+
+  @Override
+  public String getStringValue(int dictId) {
+    return _unpaddedStrings[dictId];
```
Comment (chenboat) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/OnHeapTrieBasedStringDictionary.java]: Same as above?

File: pinot-core/src/test/java/org/apache/pinot/core/segment/index/readers/ImmutableDictionaryReaderTest.java [hunk 6]
```diff
@@ -254,6 +255,15 @@ public void testOnHeapStringDictionary()
     }
   }
 
+  @Test
+  public void testOnHeapTrieBasedStringDictionary()
+      throws IOException {
+    try (OnHeapTrieBasedDictionary onHeapTrieBasedStringDictionary = new OnHeapTrieBasedStringDictionary(PinotDataBuffer.mapReadOnlyBigEndianFile(new File(TEMP_DIR, STRING_COLUMN_NAME + V1Constants.Dict.FILE_EXTENSION)), NUM_VALUES,
+        _numBytesPerStringValue, (byte) 0)) {
```
Comment (chenboat) [inline pinot-core/src/test/java/org/apache/pinot/core/segment/index/readers/ImmutableDictionaryReaderTest.java]: Add a test where paddingByte != 0

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/OnHeapTrieBasedStringDictionary.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+
+public class OnHeapTrieBasedStringDictionary extends OnHeapTrieBasedDictionary {
+  private final byte _paddingByte;
+  private final String[] _unpaddedStrings;
+  private final String[] _paddedStrings;
+  private final PatriciaTrie<Integer> _paddedStringToIdTrie;
+  private final PatriciaTrie<Integer> _unpaddedStringToIdTrie;
+
+  public OnHeapTrieBasedStringDictionary(PinotDataBuffer dataBuffer, int length, int numBytesPerValue,
```
Comment (chenboat) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/OnHeapTrieBasedStringDictionary.java]: Comments? Especially on how to use the paddingByte?

**Comment by npawar** [comment]
Closed this PR due to 6 months inactivity. Reopen if needed.

**Pull request closed without merging**